pallets/lbp/src/lib.rs

```diff
@@ -0,0 +1,964 @@
+impl<BlockNumber: AtLeast32BitUnsigned> LBPWeightCalculation<BlockNumber> for LBPWeightFunction {
+	fn calculate_weight(
+		curve_type: CurveType,
+		a_x: BlockNumber,
+		b_x: BlockNumber,
+		a_y: LBPWeight,
+		b_y: LBPWeight,
+		at: BlockNumber,
+	) -> Result<LBPWeight, MathError> {
+		match curve_type {
+			CurveType::Linear => calculate_linear_weights(a_x, b_x, a_y, b_y, at),
+			CurveType::Constant => Ok(a_y),
+		}
+	}
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::OriginFor;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// Multi currency for transfer of currencies
+		type MultiCurrency: MultiCurrencyExtended<Self::AccountId, CurrencyId = AssetId, Amount = Amount, Balance = Balance>
+			+ MultiReservableCurrency<Self::AccountId>;
+
+		#[pallet::constant]
+		/// Native Asset Id
+		type NativeAssetId: Get<AssetId>;
+
+		/// Function for calculation of LBP weights
+		type LBPWeightFunction: LBPWeightCalculation<Self::BlockNumber>;
+
+		/// Mapping of asset pairs to unique pool identities
+		type AssetPairPoolId: AssetPairPoolIdFor<AssetId, PoolId<Self>>;
+
+		#[pallet::constant]
+		/// Storage fee
+		type PoolDeposit: Get<BalanceOf<Self>>;
+
+		/// Trading fee rate
+		type ExchangeFee: Get<fee::Fee>;
+
+		/// Weight information for the extrinsics.
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Create pool errors
+		CannotCreatePoolWithSameAssets,
+		CannotCreatePoolWithZeroLiquidity,
+
+		/// Update pool errors
+		NotOwner,
+		SaleStarted,
+		SaleNotEnded,
+		InvalidData,
+		CannotPauseEndedPool,
+		CannotPausePausedPool,
+		PoolIsNotPausable,
+		CannotUnpauseEndedPool,
+		PoolIsNotPaused,
+		MaxSaleDurationExceeded,
+
+		/// Add / Remove liquidity errors
+		CannotAddZeroLiquidity,
+		CannotRemoveZeroLiquidity,
+		LiquidityOverflow,
+		LiquidityUnderflow, // TODO: do we need it? Can we use LiquidityOverflow instead?
+
+		/// Balance errors
+		InsufficientAssetBalance,
+
+		/// Pool existence errors
+		TokenPoolNotFound,
+		TokenPoolAlreadyExists,
+
+		/// Calculation errors
+		InvalidBlockNumber,
+		CalculationError, // TODO: replace me with something more meaningful?
+
+		BlockNumberInvalid,
+
+		ZeroAmount,
+		SaleIsNotRunning,
+
+		// Trading limit errors
+		MaxInRatioExceeded,
+		MaxOutRatioExceeded,
+
+		/// Invalid fee
+		FeeAmountInvalid,
+
+		// Asset limit exceeded
+		AssetBalanceLimitExceeded,
+
+		Overflow,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Create new LBP pool
+		/// who, asset a, asset b, amount_a, amount_b
+		CreatePool(T::AccountId, AssetId, AssetId, BalanceOf<T>, BalanceOf<T>),
```

Comment:
As discussed - events should be names with past tense. There is a mix of present and past tenses.